1306_expansion_statements/expansion-statements.md

```diff
@@ -469,6 +469,409 @@ void f() {
+- [[#.#]{.pnum} for an `$expansion-statement$`, its `$statement$` (but not an `$init-statement$`).]{.addu}
+
+:::
+
+Extend "enclose" to cover expansion statements in paragraph 3:
+
+::: std
+[3]{.pnum} A `$statement$` `S1` _encloses_ a `$statement$` `S2` if
+
+- [#.#]{.pnum} `S2` is a substatement of `S1`,
+- [#.#]{.pnum} `S1` is a `$selection-statement$`[,]{.addu} [or]{.rm} `$iteration-statement$`[, or `$expansion-statement$`]{.addu} and `S2` is the `$init-statement$` of `S1`,
+- [#.#]{.pnum} [...]
+
+:::
+
+## [stmt.label] Labeled statement {-}
+
+Add a new paragraph to the end of [stmt.label]:
+
+::: std
+[[4]{.pnum} An identifier label shall not occur in an `$expansion-statement$` ([stmt.expand]).]{.addu}
+
+:::
+
+## [stmt.ranged] The range-based `for` statement {-}
+
+Add the following paragraph to the end of [stmt.ranged]:
+
+::: std
+FIXME: Consider moving definition of "iterable" to [stmt.expand]
+
+[[3]{.pnum} An expression is _iterable_ if, when treated as a `$for-range-initializer$` ([stmt.iter.general]), expressions `$begin-expr$` and `$end-expr$` can be determined as specified above, and if they are of the form `begin($range$)` and `end($range$)` then argument-dependent lookup finds at least one function or function template for each.]{.addu}
+
+:::
+
+## 8.6+ [stmt.expand] Expansion statements
+
+Insert this section after [stmt.iter]{.sref} (and renumber accordingly).
+
+::: std
+::: addu
+**Expansion statements   [stmt.expand]**
+
+[1]{.pnum} Expansion statements specify repeated instantiations ([temp.spec]) of their substatement.
+
+```
+$expansion-statement$:
+    template for ( $init-statement$@~_opt_~@ $for-range-declaration$ : $expansion-initializer$ ) $statement$
+
+$expansion-initializer$:
+    $expression$
+    $expansion-init-list$
+
+$expansion-init-list$:
+    { $expression-list$ }
+```
+
+[Each `$decl-specifier$` in the `$decl-specifier-seq$` of a `$for-range-declaration$` is either a `$type-specifier$` or `constexpr` ([stmt.ranged]).]{.note}
+
+[#]{.pnum} The `$statement$` of an `$expansion-statement$` is a control-flow-limited statement ([stmt.label]).
+
+[#]{.pnum} For the purpose of name lookup and instantiation, the `$for-range-declaration$` and the `$statement$` of the `$expansion-statement$` are together considered a template definition.
+
+[#]{.pnum} An expansion statement whose `$expansion-initializer$` is an iterable expression ([stmt.ranged]) that does not have array type is an _iterable expansion statement_. An expansion statement that is not iterable and whose `$expansion-initializer$` is a non-type-dependent expression with a structured binding size ([dcl.struct.bind]) is a _destructurable expansion statement_. An expansion statement that is neither iterable nor destructurable shall have an `$expansion-initializer$` of the form `$expansion-init-list$` and is an _enumerated expansion statement_.
```

Comment:
Why not array type? I think actually this wording ends up excluding arrays entirely. 

Also "Otherwise" in the 2nd and 3rd sentence. Is there a way to bullet this? Maybe:

> An expansion statement is
> 
> * an iteratable expansion statement if [..];
> * otherwise, a destructable expansion statement if [...];
> * otherwise, an enumerated expansion statement if [...];
> * otherwise, ill-formed.

?